components/common/src/liveliness_checker.rs

```diff
@@ -0,0 +1,467 @@
+use std::{collections::HashMap,
+          sync::Mutex,
+          thread::{self,
+                   ThreadId},
+          time::{Duration,
+                 Instant}};
+
+// Threads that end normally, simply aren't tracked any longer
+enum Status {
+    Alive {
+        last_heartbeat: Instant,
+    },
+    DeadWithError {
+        time_of_death: Instant,
+        error:         String,
+    },
+}
+type NameAndStatus = (Option<String>, Status);
+type NameAndLastHeartbeat = (Option<String>, Instant);
+type NameAndErrorExitTimeAndReason<'a> = (Option<String>, Instant, &'a str);
+type ThreadStatusMap = HashMap<ThreadId, NameAndStatus>;
+lazy_static::lazy_static! {
+    static ref THREAD_STATUSES: Mutex<ThreadStatusMap> = Default::default();
+}
+
+habitat_core::env_config_duration!(ThreadAliveThreshold,
+                                       HAB_THREAD_ALIVE_THRESHOLD_SECS => from_secs,
+                                       Duration::from_secs(5 * 60));
+habitat_core::env_config_duration!(ThreadAliveCheckDelay,
+                                       HAB_THREAD_ALIVE_CHECK_DELAY_SECS => from_secs,
+                                       Duration::from_secs(60));
+habitat_core::env_config_duration!(ThreadDeadIgnoreDelay,
+                                       HAB_THREAD_DEAD_IGNORE_DELAY_SECS => from_secs,
+                                       Duration::from_secs(60 * 60 * 24 * 7)); // 1 week
+
+/// Call periodically from a thread which has a work loop to indicate that the thread is
+/// still alive and processing its loop. If this function is not called for more than
+/// `ThreadAliveThreshold`, it will be error logged as a likely deadlock.
+///
+/// If the thread exits the loop where mark_thread_alive will be called, it must be unregistered
+/// from the liveliness checker or else be subject to false positives. In general, the idiom is
+/// ```
+/// use habitat_common::liveliness_checker;
+/// # let error = false;
```

Comment:
Yes, without that line, `cargo test --doc -p habitat_common` would fail. 

The `#` is to hide it from the documentation display. See, for example: https://github.com/rust-lang/rust/blob/master/src/liballoc/string.rs#L953 vs http://doc.rust-lang.org/1.36.0/alloc/string/struct.String.html#method.try_reserve.

Reply:
Ah, gotcha... :+1: